msbrainpy/atlas/query.py

```diff
@@ -142,3 +146,165 @@ def generateQueryRows(model, criteria, options='[only$eq''genes.entrez_id,data_s
+        for row in section_data_set['msg']:
+            if row['plane_of_section_id'] == plane_of_section:
+                if age_id is None or row['reference_space']['age_id'] == age_id:
+                    dirname_section_id = 'age_id-' + str(row['reference_space']['age_id']) + '_id-' + str(row['id'])
+                    if os.path.exists(dirname_section_id):
+                        os.mkdir(dirname_section_id)
+                    out_dir = os.path.join(dirname_plane_of_section_id, dirname_section_id)
+                    func(row['id'], out_dir)
+
+
+# function for retrieving section data sets for a gene or genes
+def section_data_set_query(entrez_ids, plane_of_section=None, product_id=1, write_out=None, verbose=True):
+    criteria = section_data_set_criteria(entrez_ids, plane_of_section, product_id=product_id)
+    url_0 = "http://api.brain-map.org/api/v2/data/query.json?criteria=model::SectionDataSet,rma::criteria,"
+    url_1 = criteria
+    url_3 = ',rma::include,genes,reference_space'
+    url = url_0 + url_1 + url_3
+    response = requests.get(url)
+    sections = response.json()
+    if verbose:
+        print('{} samples matching the criteria were found'.format(sections['num_rows']))
+    if write_out is not None:
+        with open(write_out, 'w') as outfile:
+            json.dump(sections, outfile)
+        if verbose:
+            print('data was saved at {}'.format(write_out))
+    return sections
+
+
+# function for composing criteria string for experiment search
+def section_data_set_criteria(entrez_ids, plane_of_section_id=None, product_id=1):
+    """
+    FUNCTION: This returns a criteria string with which to query the allen rma api
+    ARGS:
+        :param entrez_ids: gene id or ids (int, list, None)
+        :param plane_of_section_id: 1 -> coronal, 2 -> saggital (None, int)
+        :param product_id: Mouse -> 1, default=1 (int)
+    RETURNS:
+        :return: rma criteria (str)
+
+    Example of successful criteria:
+    rma::criteria,[failed$eq'false'],products[id$eq1],plane_of_section[id$eq1],genes[entrez_id$eq14432]
+        TIP: Play with the RMA Query builder to find out about query options and variable results
+    """
+    criteria_0 = '[failed$eq\'false\'],products[id$eq{}],'.format(product_id)
+    if type(plane_of_section_id) == int:
+        criteria_1 = 'plane_of_section[id$eq{}],'.format(plane_of_section_id)
+    if plane_of_section_id is None:
+        criteria_1 = ''
+    raise_an_error = True
+    if type(entrez_ids) == int:
+        criteria_2 = 'genes[entrez_id$eq{}]'.format(entrez_ids)
+        raise_an_error = False
+    if type(entrez_ids) == list:
+        criteria_2 = 'genes[entrez_id$in{}]'.format(entrez_ids)
+        raise_an_error = False
+    if type(entrez_ids) is None:
+        criteria_2 = ''
+        raise_an_error = False
+    if raise_an_error:
+        raise ValueError('Please enter an argument for entrez_ids of type int, list, or None')

```

Comment:
This function could use some reworking:

- if `plane_of_section_id` is anything other than an int or None, you will get an UnboundLocalError because `criteria_1` will never be defined.
- rather than using the "raise_an_error" flag, you should do an elif cascade:

```python
if type(entrez_ids) == int:
    ...
elif type(entrez_ids) == list:
    ...
elif entrez_ids is None:  #  type(x) will never be None btw ;)
    ...
else:
    raise TypeError(...)
```

and it should be a TypeError rather than a valueerror since you are checking for the type of the variable, not its value.